### DISCUSSION TITLE
Can't remove stray voxels from a voxelized mesh ?

### DISCUSSION THREAD

**Discussion by SofianeHamlaoui** [pyvista/pyvista#4563] [Q&A]
So, I finally succeeded to transfer a ThreeJS BufferGeometry to Pyvista !

when I visualize the mesh I get this :

![image](https://github.com/pyvista/pyvista/assets/16967174/bab5ce82-eca4-41b6-88c3-fbbeb41cc297)

But After Voxelization I get some weird blocs on the sides :

![image](https://github.com/pyvista/pyvista/assets/16967174/b1bd1fb5-154c-460f-a898-42d723115edb)

Is there a way to fix this ?

Knowing that I have tried connectivity, lowering the value of the density on voxelisation and the issue is still there. 

Looks like this kind of issue happens when there are gaps in the buildings 

Thank you in advance.

**Comment by banesullivan** [comment]
Hm. `voxelize()` may not be the best choice here... what's your end goal with this dataset? Can you share the mesh (`mesh.save('mesh.vtk'`) so I can try to produce a better result

**Reply by SofianeHamlaoui** [reply]
Will the original .obj file I’m using to visualise and then voxelize help you ?

**Reply by banesullivan** [reply]
That works too!

**Reply by SofianeHamlaoui** [reply]
Here is it : 

https://gist.github.com/SofianeHamlaoui/167e7409a2081edac2316ed1bf4750f0

**Reply by banesullivan** [reply]
Would you clarify what your goal is? What are you working to accomplish?

**Reply by SofianeHamlaoui** [reply]
The obj file I sent you is what I should receive from the front, the goal is to get the data, voxelize it and send it back to the font as an .obj file.

the issue here is that when I visualize the original mesh it looks just perfect, but it gets some weird stray voxels when I try to voxelize it using Pyvista! 

As I said, this issue happens when the buildings in the original mesh file has gaps…

**Reply by banesullivan** [reply]
> voxelize it and send it back to the font as an .obj file.

Hm... I may be wrong, but I don't think the OBJ format supports voxel (or any 3D) cell types. My impression is that it is vertex/line/face only. (But I could be wrong!) So we may need to use a different export format after voxelizing.

> the issue here is that when I visualize the original mesh it looks just perfect, but it gets some weird stray voxels when I try to voxelize it using Pyvista!

I'll see if I can manage to get anything cleaner... fingers crossed

**Reply by SofianeHamlaoui** [reply]
Okey, thank you a lot.

You can always check the voxelization using this website to preview it : 

https://drububu.com/miscellaneous/voxelizer/?out=obj ( and it works perfectly on this website btw ) 

Also, when I try voxelization of buildings without gaps from the same source ( a Three.Js preview map ) it works perfectly.

For the export, I did try to export it to obj, then used an online obj previewer and I had the same stray voxels, that’s how I understood that the issue comes from the voxelization using Pyvista itself.

**Reply by banesullivan** [reply]
The artifacts ar because the surface is non-manifold. We'll have to take a different approach to this. Also, how do you want to handle thin surfaces like the following:
![screenshot (16)](https://github.com/pyvista/pyvista/assets/22067021/f1408e2d-1601-4788-b1fc-e3a0fe552efb)

This is a tough mesh to "voxelize"

**Reply by banesullivan** [reply]
> https://drububu.com/miscellaneous/voxelizer/?out=obj ( and it works perfectly on this website btw )

This website is super cool. Yeah. Woah. This might be hard to beat.

**Reply by SofianeHamlaoui** [reply]
> The artifacts ar because the surface is non-manifold. We'll have to take a different approach to this. Also, how do you want to handle thin surfaces like the following:

So.. yeah I thought about filling the gaps on the original mesh before the voxelization.. didn't succeed. 

But we don't have too much buildings with this issue, as it's a kind of formal/gov data that's the only true, realistic and up to date data of the buildings in that area.. so we have to deal with it 


The only solutions I'm seeing now is : 
 * To fill the gaps before the Voxelization 
 *  Use another export format so it can probably work ( even if I don't think this is the right solution to have )

**Reply by SofianeHamlaoui** [reply]
> This website is super cool. Yeah. Woah. This might be hard to beat. 

Haha yeah, the website is just to show that the voxelization works just perfectly on that mesh... I use it for testing

**Comment by SofianeHamlaoui** [comment]
Closing after trying everything possible to do and not working, due to the quality of the received meshs I want to voxelize!
